lib/runtime/wasmer/exports.go

```diff
@@ -227,6 +229,53 @@ func (in *Instance) QueryCallFeeDetails(ext []byte) (*types.FeeDetails, error) {
 // TODO: use this in block verification process (#1873)
 func (in *Instance) CheckInherents() {}
 
+// GrandpaGenerateKeyOwnershipProof returns grandpa key ownership proof from runtime.
+func (in *Instance) GrandpaGenerateKeyOwnershipProof(authSetID uint64, authorityID ed25519.PublicKeyBytes) (
+	types.OpaqueKeyOwnershipProof, error) {
+
+	const bufferSize = 8 + 32 // authSetID uint64 + ed25519.PublicKeyBytes
+	buffer := bytes.NewBuffer(make([]byte, 0, bufferSize))
+	encoder := scale.NewEncoder(buffer)
+	err := encoder.Encode(authSetID)
+	if err != nil {
+		return nil, fmt.Errorf("encoding auth set id: %w", err)
+	}
+	err = encoder.Encode(authorityID)
+	if err != nil {
+		return nil, fmt.Errorf("encoding authority id: %w", err)
+	}
+	ret, err := in.Exec(runtime.GrandpaGenerateKeyOwnershipProof, buffer.Bytes())
+	if err != nil {
+		return nil, err
+	}
+
+	keyOwnershipProof := types.OpaqueKeyOwnershipProof{}
+	err = scale.Unmarshal(ret, &keyOwnershipProof)
+	if err != nil {
+		return nil, fmt.Errorf("scale decoding: %w", err)
+	}
+
+	return keyOwnershipProof, nil
+}
+
+// GrandpaSubmitReportEquivocationUnsignedExtrinsic reports equivocation report to the runtime.
+func (in *Instance) GrandpaSubmitReportEquivocationUnsignedExtrinsic(
+	equivocationProof types.GrandpaEquivocationProof, keyOwnershipProof types.OpaqueKeyOwnershipProof,
+) error {
+	buffer := bytes.NewBuffer(nil)
+	encoder := scale.NewEncoder(buffer)
+	err := encoder.Encode(equivocationProof)
+	if err != nil {
+		return fmt.Errorf("encoding equivocation proof: %w", err)
+	}
+	err = encoder.Encode(keyOwnershipProof)
+	if err != nil {
+		return fmt.Errorf("encoding key ownership proof: %w", err)
+	}
+	_, err = in.Exec(runtime.GrandpaSubmitReportEquivocation, buffer.Bytes())
+	return err
```

Comment:
nit have it as

```suggestion
	if err != nil {
		return err
	}
	
	return nil
```

so that you can more easily spot code not covered. I would guess the exec error case is not test covered 🤔 Would it be possible to plug in bad equivocationProof + keyOwnershipProof to make the runtime return an error?

Reply:
Commented above as well, but it appears to just return an empty proof when provided with incorrect (but correctly encoded) data

Reply:
Similar to the above case, it doesn't seem to error unless due to formatting